core/src/main/scala/com/banno/kafka/consumer/ConsumerOps.scala

```diff
@@ -391,100 +418,164 @@ case class ConsumerOps[F[_], K, V](consumer: ConsumerApi[F, K, V]) {
-              () <- s
-                .needToCommit(maxRecordCount, now, maxElapsedTime)
-                .traverse_(os =>
-                  commit(os) *>
-                  state.update(_.reset(now))
-                )
+              b <- process(a).onError(_ =>
+                offsetChannel.close.void
+              ) // Signal to flush the commit stream
+              _ <- offsetChannel
+                .send(offsets0)
+                .replicateA_(num) // Stuff the queue to trigger the barrier
             } yield b
           }
-          .onFinalize(commitNextOffsets)
+
+          def doCommit(chunk: Chunk[Map[TopicPartition, Long]]): F[Unit] =
+            F.uncancelable { _ =>
+              (if (chunk.isEmpty) ().pure[F]
+               else {
+                 val offsets =
+                   chunk.foldLeft(Map.empty[TopicPartition, Long])(_ ++ _)
+                 keepAlive.commit(offsets)
+               })
+            }
+
+          val commitStream =
+            offsetChannel.stream
+              .groupWithin(maxRecordCount0, maxElapsedTime)
+              .evalMap(doCommit)
+
+          val processStream =
+            stream
+              .evalMap(processAndEnqueueOffsets)
+              .onFinalize(
+                offsetChannel.close.void
+              ) // Signal to flush the commit stream
+
+          Stream
+            .bracket(commitStream.compile.drain.start)(_.join.void)
+            .flatMap(_ => processStream)
+        }
+    }
+  }
+
+  private sealed trait KeepAlive {
+    def commit(offsets: Map[TopicPartition, Long]): F[Unit]
+    def stream[A](gs: KeepAlive => Stream[F, A]): Stream[F, A]
+  }
+
+  private object KeepAlive {
+    def apply[A](interval: Duration)(
+        stream: KeepAlive => Stream[F, A]
+    )(implicit F: Temporal[F]): Stream[F, A] =
+      interval match {
+        case interval0: FiniteDuration =>
+          val init: F[KeepAlive] =
+            for {
+              assignments <- consumer.assignment
+              offsets <- consumer.partitionQueries.committed(
+                assignments
+              )
+              keepAlive <-
+                if (assignments.isEmpty) NoKeepAlive.pure[F]
+                else
+                  for {
+                    lock <- Semaphore(1)
+                    ref <- Ref.of(
+                      offsets.view.mapValues(_.offset).toMap
+                    )
+                  } yield HasKeepAlive(interval0, lock, ref)
+            } yield keepAlive
+
+          Stream.eval(init).flatMap(_.stream(stream))
+        case _ => stream(NoKeepAlive)
       }
 
-  case class OffsetCommitState(
-      offsets: Map[TopicPartition, Long],
-      recordCount: Long,
-      lastCommitTime: FiniteDuration,
-  ) {
-    def update(os: Map[TopicPartition, Long], count: Long): OffsetCommitState =
-      copy(
-        offsets = offsets ++ os,
-        recordCount = recordCount + count,
-      )
+    case object NoKeepAlive extends KeepAlive {
```

Comment:
The `KeepAlive` trait is private, so should `NoKeepAlive` also be explicitly made private?

Reply:
`NoKeepAlive` and `HasKeepAlive` are nested in the private `KeepAlive` object so they're already inaccessible. Happy to change, but it's redundant.

Reply:
I missed that they were nested. Ignore me 🙂